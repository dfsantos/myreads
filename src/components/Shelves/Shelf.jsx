import React from 'react';
import PropTypes from 'prop-types';
import Paper from 'material-ui/Paper';

import Book from '../Book';

const propTypes = {
  books: PropTypes.array,
  name: PropTypes.string.isRequired,
  shelf: PropTypes.string.isRequired,
  onCategorizeBook: PropTypes.func.isRequired,
};

const defaultProps = {
  books: [],
};

const byShelf = shelf => book => book.shelf === shelf;

function Shelf({ name, shelf, books, onCategorizeBook }) {
  const shelfBooks = books.filter(byShelf(shelf));
  const isEmpty = shelfBooks.length === 0;
  return (
    <Paper className="shelf">
      <div className="shelf-title">
        <span>{name}</span>
      </div>
      <div className="shelf-content">
        {isEmpty && (
          <div className="empty-shelf-message">
            <span>This shelf is empty.</span>
          </div>
        )}
        {!isEmpty &&
          shelfBooks.map(book => {
            const { thumbnail } = book.imageLinks || { thumbnail: '' };
            return (
              <Book
                key={book.id}
                title={book.title}
                subtitle={book.subtitle}
                coverLink={thumbnail}
                authors={book.authors}
                description={book.description}
                shelf={book.shelf}
                onChangeBookShelf={onCategorizeBook(book)}
              />
            );
          })}
      </div>
    </Paper>
  );
}

Shelf.propTypes = propTypes;
Shelf.defaultProps = defaultProps;

export default Shelf;
