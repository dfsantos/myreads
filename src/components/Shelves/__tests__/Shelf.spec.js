import React from 'react';
import toJson from 'enzyme-to-json';

import Shelf from '../Shelf';

const BOOKSHELF_NAME = 'Shelf Name';
const BOOKSHELF_CATEGORY = 'read';

const onChangeBookCategory = jest.fn(() => () => {});

it('should render without erros', () => {
  shallow(
    <Shelf
      name={BOOKSHELF_NAME}
      shelf={BOOKSHELF_CATEGORY}
      onCategorizeBook={onChangeBookCategory}
    />
  );
});

it('snapshot without books', () => {
  const wrapper = shallow(
    <Shelf
      name={BOOKSHELF_NAME}
      shelf={BOOKSHELF_CATEGORY}
      onCategorizeBook={onChangeBookCategory}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('snapshot with books', () => {
  const books = [
    { id: 1, title: 'Book A', shelf: BOOKSHELF_CATEGORY, imageLinks: { thumbnail: '' } },
    { id: 2, title: 'Book B', shelf: BOOKSHELF_CATEGORY, imageLinks: { thumbnail: '' } },
  ];
  const wrapper = shallow(
    <Shelf
      books={books}
      name={BOOKSHELF_NAME}
      shelf={BOOKSHELF_CATEGORY}
      onCategorizeBook={onChangeBookCategory}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('onChangeBookCategory should be called with book as parameter', () => {
  const book = {
    id: 1,
    title: 'Book A',
    shelf: BOOKSHELF_CATEGORY,
    imageLinks: { thumbnail: '' },
  };
  const wrapper = shallow(
    <Shelf
      books={[book]}
      name={BOOKSHELF_NAME}
      shelf={BOOKSHELF_CATEGORY}
      onCategorizeBook={onChangeBookCategory}
    />
  );
  expect(onChangeBookCategory).toHaveBeenCalledWith(book);
});

it('should filter book by shelf category', () => {
  const bookA = {
    id: 1,
    title: 'Book A',
    shelf: BOOKSHELF_CATEGORY,
    imageLinks: { thumbnail: '' },
  };
  const bookB = {
    id: 2,
    title: 'Book B',
    shelf: 'wrongCategory',
    imageLinks: { thumbnail: '' },
  };
  const wrapper = shallow(
    <Shelf
      books={[bookA, bookB]}
      name={BOOKSHELF_NAME}
      shelf={BOOKSHELF_CATEGORY}
      onCategorizeBook={onChangeBookCategory}
    />
  );
  expect(onChangeBookCategory).toHaveBeenCalledWith(bookA);
  expect(onChangeBookCategory).not.toHaveBeenCalledWith(bookB);
});
