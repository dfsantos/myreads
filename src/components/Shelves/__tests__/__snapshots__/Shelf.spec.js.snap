// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot with books 1`] = `
<Paper
  circle={false}
  className="shelf"
  rounded={true}
  transitionEnabled={true}
  zDepth={1}
>
  <div
    className="shelf-title"
  >
    <span>
      Shelf Name
    </span>
  </div>
  <div
    className="shelf-content"
  >
    <Book
      authors={Array []}
      coverLink=""
      key="1"
      onChangeBookShelf={[Function]}
      shelf="read"
      title="Book A"
    />
    <Book
      authors={Array []}
      coverLink=""
      key="2"
      onChangeBookShelf={[Function]}
      shelf="read"
      title="Book B"
    />
  </div>
</Paper>
`;

exports[`snapshot without books 1`] = `
<Paper
  circle={false}
  className="shelf"
  rounded={true}
  transitionEnabled={true}
  zDepth={1}
>
  <div
    className="shelf-title"
  >
    <span>
      Shelf Name
    </span>
  </div>
  <div
    className="shelf-content"
  >
    <div
      className="empty-shelf-message"
    >
      <span>
        This shelf is empty.
      </span>
    </div>
  </div>
</Paper>
`;
